circonusllhist.go

```diff
@@ -562,32 +562,40 @@ func (h *Histogram) insertBin(hb *bin, count int64) uint64 {
 	}
 	found, idx := h.internalFind(hb)
 	if !found {
-		if h.used == h.allocd {
-			new_bvs := make([]bin, h.allocd+defaultHistSize)
-			if idx > 0 {
-				copy(new_bvs[0:], h.bvs[0:idx])
-			}
-			if idx < h.used {
-				copy(new_bvs[idx+1:], h.bvs[idx:])
-			}
-			h.allocd = h.allocd + defaultHistSize
-			h.bvs = new_bvs
-		} else {
-			copy(h.bvs[idx+1:], h.bvs[idx:h.used])
-		}
-		h.bvs[idx].val = hb.val
-		h.bvs[idx].exp = hb.exp
-		h.bvs[idx].count = uint64(count)
-		h.used++
-		for i := idx; i < h.used; i++ {
-			f2 := h.bvs[i].newFastL2()
-			if h.lookup[f2.l1] == nil {
-				h.lookup[f2.l1] = make([]uint16, 256)
-			}
-			h.lookup[f2.l1][f2.l2] = uint16(i) + 1
+		return h.insertNewBinAt(idx, hb, count)
+	}
+	return h.updateOldBinAt(idx, hb, count)
+}
+
+func (h *Histogram) insertNewBinAt(idx uint16, hb *bin, count int64) uint64 {
+	if h.used == h.allocd {
+		new_bvs := make([]bin, h.allocd+defaultHistSize)
+		if idx > 0 {
+			copy(new_bvs[0:], h.bvs[0:idx])
+		}
+		if idx < h.used {
+			copy(new_bvs[idx+1:], h.bvs[idx:])
 		}
-		return h.bvs[idx].count
+		h.allocd = h.allocd + defaultHistSize
+		h.bvs = new_bvs
+	} else {
+		copy(h.bvs[idx+1:], h.bvs[idx:h.used])
+	}
+	h.bvs[idx].val = hb.val
+	h.bvs[idx].exp = hb.exp
+	h.bvs[idx].count = uint64(count)
+	h.used++
+	for i := idx; i < h.used; i++ {
```

Comment:
For a large merge, we'll end up redoing the fast look indexes A LOT.

Since this insertat and updateold function are internal, we should create a updateFast function.

We hold the locks already, so merge them all (call all the update/inserts) then call updateFast once.

The performance difference will be very large.  It might be worth writing a performance test as well (so you can see the difference).

Reply:
@postwait It does improve the performance, here is the changes between benchmarks.

```sh
$ go test -v -run=none -bench=Merge -benchtime=5s .
$ benchcmp old.txt new.txt
benchmark                                  old ns/op     new ns/op     delta
BenchmarkHistogramMerge/random-4           165594        143038        -13.62%
BenchmarkHistogramMerge/large_insert-4     78160         74364         -4.86%

benchmark                                  old allocs     new allocs     delta
BenchmarkHistogramMerge/random-4           11             11             +0.00%
BenchmarkHistogramMerge/large_insert-4     12             12             +0.00%

benchmark                                  old bytes     new bytes     delta
BenchmarkHistogramMerge/random-4           26108         26109         +0.00%
BenchmarkHistogramMerge/large_insert-4     26112         26112         +0.00%
```

The source code of benchmark is [here](https://github.com/circonus-labs/circonusllhist/pull/14/files#diff-095e965c1e92c9277036371d89bd083cR301) .